### ISSUE TITLE
Problems with dnsmasq and dnscrypt-proxy, request times out

### ISSUE THREAD

**Issue by shibumi** [jedisct1/dnscrypt-proxy#468]
Hello,
I have tried to setup dnscrypt-proxy with dnsmasq as dns cache. But I don't get a reply for a DNS request. It seems like the forwarding does from dnscrypt to dnsmasq does not work and sometimes dnscrypt does not work. With `tcpdump` I can see requests to the dnscrypt server, but I just get 'reset' packages back?! And I can see that dnsmasq gets the dns requests but there is no forward from dnscrypt to dnsmasq.

Here is my configuration:

/etc/dnsmasq.conf
```
proxy-dnssec
no-resolv
server=127.0.0.1#40
listen-address=127.0.0.1
```

/etc/systemd/system/dnscrypt-proxy.service
```
# /usr/lib/systemd/system/dnscrypt-proxy.service
[Unit]
Description=DNSCrypt client proxy
Documentation=man:dnscrypt-proxy(8)
Requires=dnscrypt-proxy.socket
After=network.target
Before=nss-lookup.target

[Install]
Also=dnscrypt-proxy.socket
WantedBy=multi-user.target

[Service]
Type=simple
NonBlocking=true

# Fill in the resolver name with one from dnscrypt-resolvers.csv file
# It is also recommended to create a dedicated system user, for example _dnscrypt
# Additional features, such as ephemeral keys and plugins, can be enabled here as well
ExecStart=/usr/bin/dnscrypt-proxy \
          --resolver-name=<resolver name> \
          --user=<user name to run the service as>

# /etc/systemd/system/dnscrypt-proxy.service.d/override.conf
[Service]
ExecStart=
ExecStart=/usr/bin/dnscrypt-proxy \
          --resolver-name=dnscrypt.eu-nl\
          --user=chris
```

/etc/systemd/systemd/dnscrypt-proxy.socket
```

# /usr/lib/systemd/system/dnscrypt-proxy.socket
[Unit]
Description=dnscrypt-proxy listening socket

[Socket]
ListenStream=127.0.0.1:53
ListenDatagram=127.0.0.1:53

[Install]
WantedBy=sockets.target

# /etc/systemd/system/dnscrypt-proxy.socket.d/override.conf
[Socket]
ListenStream=
ListenDatagram=
ListenStream=127.0.0.1:40
ListenDatagram=127.0.0.1:40
```

systemctl status dnscrypt-proxy:
```
● dnscrypt-proxy.service - DNSCrypt client proxy
   Loaded: loaded (/usr/lib/systemd/system/dnscrypt-proxy.service; enabled; vendor preset: disabled)
  Drop-In: /etc/systemd/system/dnscrypt-proxy.service.d
           └─override.conf
   Active: active (running) since Sun 2016-09-25 17:28:34 CEST; 7h ago
     Docs: man:dnscrypt-proxy(8)
 Main PID: 509 (dnscrypt-proxy)
    Tasks: 1 (limit: 4915)
   CGroup: /system.slice/dnscrypt-proxy.service
           └─509 /usr/bin/dnscrypt-proxy --resolver-name=dnscrypt.eu-nl --user=chris

Sep 25 23:33:45 motoko dnscrypt-proxy[509]: [INFO] Refetching server certificates
Sep 25 23:33:45 motoko dnscrypt-proxy[509]: [INFO] Server certificate with serial '0001' received
Sep 25 23:33:45 motoko dnscrypt-proxy[509]: [INFO] This certificate is valid
Sep 25 23:33:45 motoko dnscrypt-proxy[509]: [INFO] Chosen certificate #808464433 is valid from [2016-09-08] to [2017-09-08]
Sep 25 23:33:45 motoko dnscrypt-proxy[509]: [INFO] Server key fingerprint is 72DF:BE14:531F:F2AD:FD0F:BC8B:F711:B93D:799F:E4D0:34EC:D26B:8BF9:FFA9:32E7:2B79
Sep 26 00:34:36 motoko dnscrypt-proxy[509]: [INFO] Refetching server certificates
Sep 26 00:34:36 motoko dnscrypt-proxy[509]: [INFO] Server certificate with serial '0001' received
Sep 26 00:34:36 motoko dnscrypt-proxy[509]: [INFO] This certificate is valid
Sep 26 00:34:36 motoko dnscrypt-proxy[509]: [INFO] Chosen certificate #808464433 is valid from [2016-09-08] to [2017-09-08]
Sep 26 00:34:36 motoko dnscrypt-proxy[509]: [INFO] Server key fingerprint is 72DF:BE14:531F:F2AD:FD0F:BC8B:F711:B93D:799F:E4D0:34EC:D26B:8BF9:FFA9:32E7:2B79
```

systemctl status dnsmasq.service
```
● dnsmasq.service - A lightweight DHCP and caching DNS server
   Loaded: loaded (/usr/lib/systemd/system/dnsmasq.service; enabled; vendor preset: disabled)
   Active: active (running) since Mon 2016-09-26 00:51:26 CEST; 15min ago
     Docs: man:dnsmasq(8)
 Main PID: 20595 (dnsmasq)
    Tasks: 1 (limit: 4915)
   CGroup: /system.slice/dnsmasq.service
           └─20595 /usr/bin/dnsmasq -k --enable-dbus --user=dnsmasq --pid-file

Sep 26 00:51:26 motoko systemd[1]: Starting A lightweight DHCP and caching DNS server...
Sep 26 00:51:26 motoko dnsmasq[20593]: dnsmasq: syntax check OK.
Sep 26 00:51:26 motoko dnsmasq[20595]: started, version 2.76 cachesize 150
Sep 26 00:51:26 motoko dnsmasq[20595]: compile time options: IPv6 GNU-getopt DBus i18n IDN DHCP DHCPv6 no-Lua TFTP conntrack ipset auth DNSSEC loop-detect inotify
Sep 26 00:51:26 motoko dnsmasq[20595]: DBus support enabled: connected to system bus
Sep 26 00:51:26 motoko dnsmasq[20595]: using nameserver 127.0.0.1#40
Sep 26 00:51:26 motoko systemd[1]: Started A lightweight DHCP and caching DNS server.
Sep 26 00:51:26 motoko dnsmasq[20595]: read /etc/hosts - 11843 addresses
```

**Comment by jedisct1** [comment]
Does it work if you directly send the queries to the dnscrypt proxy?

```
$ hostip -r 127.0.0.1:40 example.com
```

I'm not familiar with dnsmasq, but it looks like the following is required in `dnsmasq.conf` to properly forward signed responses:

```
conf-file=/usr/share/dnsmasq/trust-anchors.conf
dnssec
```

**Comment by shibumi** [comment]
Hello @jedisct1 
I have tried it today again. And I don't know why but it seems like it's working now. I haven't changed anything..

How can I check if dnscrypt is really working? I did a dnsleak and that gave me the right dns-server back.
I think this looks fine now..

```
04:17:32.792678 IP motoko.38731 > resolver2.dnscrypt.eu.https: UDP, length 512
04:17:32.848068 IP motoko.38731 > resolver2.dnscrypt.eu.https: UDP, length 512
04:17:32.848849 IP motoko.38731 > resolver2.dnscrypt.eu.https: UDP, length 512
04:17:32.904585 IP motoko.38731 > resolver2.dnscrypt.eu.https: UDP, length 512
04:17:32.907033 IP motoko.38731 > resolver2.dnscrypt.eu.https: UDP, length 512
```

But I have one strange thing. When i try to capture traffic with wireshark and capture filter `dns` I get nothing. Is this normal? As far as I can see it should be, because dnscrypt seems to speak with the dns-server over port 443...  But normally I should see at least the traffic on the loopback device or am I wrong?

**Comment by jedisct1** [comment]
Looks good!

You can resolve `resolver.dnscrypt.org` to confirm the IP address of the resolver being used:

```bash
hostip resolver.dnscrypt.org
```

Or you can stop the proxy, and if you can't resolve anything any more, it means that you were using it :)

You should certainly see outgoing traffic. Try different host names (so they are not cached by dnsmasq) and make sure that you are looking at the external interface.

**Comment by shibumi** [comment]
Thanks. `hostip` gives me the right IP-Adress. But I still don't see any DNS traffic with wireshark. Even when I try absolutly new hostnames that can't be in the cache.. But when I check the traffic to the resolver I can see 'malformed packets' so this should work now. But still weird that I can't see any DNS traffic... 
I guess it's because dnscrypt uses port 433 not 53.

**Comment by shibumi** [comment]
Hello,
OK. Sorry. I didn't check the IP that I got from `hostip`. Why do I get 74.125.73.72 as answer for resolver.dnscrypt.org? The IP refers to Google. seems like `hostip` uses the google DNS server as default oO. It's the only tool where I can log dns requests with wireshark. When I do for example: `dig resolver.dnscrypt.org` everything is fine. Or when I try something like: `ping domain-that-is-definitly-not-in-cache.com` I also don't get dns request/replies in wireshark.

**Comment by jedisct1** [comment]
My bad, `hostip` uses Google resolvers by default. You need to add `-r 127.0.0.1` in order to send the query to `localhost` :)

**Comment by shibumi** [comment]
Yeah, I know. It's everything fine now.
I did it totally different. I am using `systemd-resolved` as DNS stub resolver now. It's working fine! No need for `dnsmasq` anymore. In Case you are interested in my setup: https://wiki.archlinux.org/index.php/DNSCrypt#Example:_configuration_for_systemd-resolved

**Comment by jedisct1** [comment]
Glad to hear that everything's working now!

**Comment by Salothin** [comment]
I think this is still an issue, I am using dnscrypt-proxy 1.9.4 and dnsmasq 2.76.

dnscrypt-proxy alone works OK
dnscrypt-proxy with systemd-resolved works OK

But with dnsmasq, dnsmasq times out.

dnsmasq log show something like this:

> server 127.0.2.1#53: queries sent 9, retried or failed 2

If I put a remote resolver like 8.8.8.8 in dnsmasq config, dnsmasq resolves but using 8.8.8.8 not the dnscrypt-proxy one.

I've tried the tcp only switch on dnscrypt-proxy, it didn't help.

Is this related to the issue #587 ?
